wolk/service/firmware_update/debian/DebianPackageInstaller.cpp

```diff
@@ -0,0 +1,187 @@
+            return;
+        }
+        callback(pathToDebianFile, false);
+    }};
+
+    // Attempt to install the package with APT
+    return m_aptPackageInstaller->installPackage(pathToDebianFile, internalCallback) == InstallationResult::Installing;
+}
+
+InstallResponse DebianPackageInstaller::installFirmware(const std::string& deviceKey, const std::string& fileName)
+{
+    // We should do this synchronously. Prepare things to be set.
+    auto conditionVariable = std::make_shared<std::condition_variable>();
+    auto mutex = std::make_shared<std::mutex>();
+    {
+        std::lock_guard<std::mutex> lockGuard{m_mapMutex};
+        m_devicesInstallingFiles[deviceKey] = fileName;
+        m_deviceInstallationResult[deviceKey] = InstallResponse::WILL_INSTALL;
+        m_conditionVariables[fileName] = conditionVariable;
+    }
+
+    // Run the update
+    const auto mutexWeakPtr = std::weak_ptr<std::mutex>{mutex};
+    auto installing = update(FileSystemUtils::absolutePath(fileName), [this, deviceKey, fileName, mutexWeakPtr](
+                                                                        const std::string&, bool installation) {
+        if (auto mutexValue = mutexWeakPtr.lock())
+        {
+            std::lock_guard<std::mutex> lockGuard{*mutexValue};
+            auto result = m_deviceInstallationResult.find(deviceKey);
+            if (result != m_deviceInstallationResult.cend())
+                result->second = installation ? InstallResponse::INSTALLED : InstallResponse::FAILED_TO_INSTALL;
+        }
+        const auto cvIt = m_conditionVariables.find(fileName);
+        if (cvIt != m_conditionVariables.cend())
+            cvIt->second->notify_one();
+    });
+
+    // Wait for the installation to be done
+    if (installing)
+    {
+        std::unique_lock<std::mutex> lock{*mutex};
+        conditionVariable->wait_for(lock, std::chrono::seconds{120});
+    }
+    else
+    {
+        m_deviceInstallationResult[deviceKey] = InstallResponse::FAILED_TO_INSTALL;
+    }
+    // Remove the values from the maps
+    {
+        std::lock_guard<std::mutex> lock{m_mapMutex};
+        m_devicesInstallingFiles.erase(deviceKey);
+        m_conditionVariables.erase(fileName);
+    }
+    return m_deviceInstallationResult[deviceKey];
+}
+
+void DebianPackageInstaller::abortFirmwareInstall(const std::string& deviceKey)
+{
+    // Check if there is a condition variable waiting for the device
+    std::lock_guard<std::mutex> lock{m_mapMutex};
+    const auto deviceIt = m_devicesInstallingFiles.find(deviceKey);
+    if (deviceIt != m_devicesInstallingFiles.cend())
+    {
+        const auto cvIt = m_conditionVariables.find(deviceIt->second);
+        if (cvIt != m_conditionVariables.cend())
+        {
+            cvIt->second->notify_one();
+        }
+    }
+}
+
+bool DebianPackageInstaller::wasFirmwareInstallSuccessful(const std::string&, const std::string&)
+{
+    return true;
```

Comment:
We don't have anything else to report. The software versions can be either the same or newer, so we can't use the default comparison. Just the fact that the software has started is an information that the restart has been successful. And since the APT installation was already successful (and the systemd restart wouldn't happen if the APT installation wasn't), there's nothing else to conclude other than the installation was successful.